lib/utils.dart

```diff
@@ -47,6 +52,20 @@ void hideLoginModal() {
   loginModal.setAttribute('hidden', 'true');
 }
 
+void setLoginDomains(List<String> domains) {
```

Comment:
(I'm leaving this comment here, as I can't comment on a file itself)

I think I may have overlooked the name of this file in the previous PR, thinking it was some kind of view. It's currently doing too many disparate things, which could be moved into other files as per the groups you're already finding yourself creating here.

Conceptually to me, the routing functions above feel like they should hang around the same code as the nav-related code, and the loading indicators, modals, login buttons and errors would fit in a more general app view.

I think `utils` as a concept should be used sparingly, and only for code that really doesn't fit anywhere else sensibly. E.g. a sorting algorithm, generating IDs, something like that. It's a slippery concept, and far too general for its own good - from a certain angle, anything can be a utility function! And particularly for UI code, things that modify the HTML should probably be somewhere on the view side of the architecture, not in utils.

Hope this makes sense, but if not, happy to chat more on Slack. Feel free to also address this in a separate PR.